File: banco_dados/utils/cache.py
"""
Sistema de Cache - Dashboard-TRONIK
===================================
Cache simples em memória para dados estáticos.
"""

from typing import Dict, Optional, Any, Callable
from datetime import datetime, timedelta
import threading
import logging

logger = logging.getLogger(__name__)


class CacheMemoria:
    """Cache simples em memória com TTL"""
    
    def __init__(self):
        self._cache: Dict[str, Dict[str, Any]] = {}
        self._lock = threading.Lock()
    
    def obter(self, chave: str, ttl_segundos: int = 3600) -> Optional[Any]:
        """
        Obtém valor do cache se ainda válido.
        
        Args:
            chave: Chave do cache
            ttl_segundos: Tempo de vida em segundos (padrão: 1 hora)
        
        Returns:
            Valor do cache ou None se expirado/não existe
        """
        with self._lock:
            if chave not in self._cache:
                return None
            
            entrada = self._cache[chave]
            expiracao = entrada['criado_em'] + timedelta(seconds=ttl_segundos)
            
            if datetime.now() > expiracao:
                # Expirado - remover
                del self._cache[chave]
                return None
            
            return entrada['valor']
    
    def definir(self, chave: str, valor: Any) -> None:
        """
        Define valor no cache.
        
        Args:
            chave: Chave do cache
            valor: Valor a armazenar
        """
        with self._lock:
            self._cache[chave] = {
                'valor': valor,
                'criado_em': datetime.now()
            }
    
    def invalidar(self, chave: str) -> None:
        """
        Remove entrada do cache.
        
        Args:
            chave: Chave do cache
        """
        with self._lock:
            if chave in self._cache:
                del self._cache[chave]
    
    def limpar(self) -> None:
        """Limpa todo o cache"""
        with self._lock:
            self._cache.clear()
    
    def obter_ou_calcular(
        self,
        chave: str,
        calcular_fn: Callable[[], Any],
        ttl_segundos: int = 3600
    ) -> Any:
        """
        Obtém do cache ou calcula e armazena.
        
        Args:
            chave: Chave do cache
            calcular_fn: Função para calcular valor se não estiver no cache
            ttl_segundos: Tempo de vida em segundos
        
        Returns:
            Valor do cache ou calculado
        """
        valor = self.obter(chave, ttl_segundos)
        if valor is not None:
            return valor
        
        # Calcular e armazenar
        valor = calcular_fn()
        self.definir(chave, valor)
        return valor


# Instância global do cache
_cache_global = CacheMemoria()


def obter_cache() -> CacheMemoria:
    """Obtém instância global do cache"""
    return _cache_global

File: estatico/js/utils/cache.js
/**
 * Sistema de Cache - Dashboard-TRONIK (Frontend)
 * ================================================
 * Cache simples em memória para requisições GET.
 */

const CacheFrontend = (function() {
    'use strict';
    
    const cache = new Map();
    const pending = new Map();
    const TTL_PADRAO = 5 * 60 * 1000; // 5 minutos
    
    /**
     * Obtém valor do cache
     */
    function obter(chave) {
        const entrada = cache.get(chave);
        if (!entrada) {
            return null;
        }
        
        // Verificar se expirou
        if (Date.now() > entrada.expiracao) {
            cache.delete(chave);
            return null;
        }
        
        return entrada.valor;
    }
    
    /**
     * Define valor no cache
     * Trata quota exceeded do localStorage
     */
    function definir(chave, valor, ttl = TTL_PADRAO) {
        try {
            cache.set(chave, {
                valor: valor,
                expiracao: Date.now() + ttl
            });
        } catch (e) {
            if (e.name === 'QuotaExceededError') {
                // Limpar cache antigo (remover 50% das entradas mais antigas)
                limparCacheAntigo();
                // Tentar novamente
                try {
                    cache.set(chave, {
                        valor: valor,
                        expiracao: Date.now() + ttl
                    });
                } catch (e2) {
                    console.warn('Cache não pôde ser salvo após limpeza:', e2);
                }
            } else {
                throw e;
            }
        }
    }
    
    /**
     * Limpa cache antigo (50% das entradas mais antigas)
     */
    function limparCacheAntigo() {
        const entradas = Array.from(cache.entries());
        if (entradas.length === 0) return;
        
        // Ordenar por expiração (mais antigas primeiro)
        entradas.sort((a, b) => a[1].expiracao - b[1].expiracao);
        
        // Remover 50% das mais antigas
        const quantidadeRemover = Math.floor(entradas.length / 2);
        for (let i = 0; i < quantidadeRemover; i++) {
            cache.delete(entradas[i][0]);
        }
    }
    
    /**
     * Remove entrada do cache
     */
    function invalidar(chave) {
        cache.delete(chave);
    }
    
    /**
     * Limpa todo o cache
     */
    function limpar() {
        cache.clear();
    }
    
    /**
     * Obtém ou calcula valor (padrão de cache)
     */
    async function obterOuCalcular(chave, calcularFn, ttl = TTL_PADRAO) {
        const valor = obter(chave);
        if (valor !== null) {
            return valor;
        }
        if (pending.has(chave)) {
            return pending.get(chave);
        }
        const promessa = (async function () {
            try {
                const novoValor = await calcularFn();
                definir(chave, novoValor, ttl);
                return novoValor;
            } finally {
                pending.delete(chave);
            }
        })();
        pending.set(chave, promessa);
        return promessa;
    }
    
    return {
        obter,
        definir,
        invalidar,
        limpar,
        obterOuCalcular
    };
})();

// Exportar globalmente
window.CacheFrontend = CacheFrontend;

